fly/integration/builds_test.go

```diff
@@ -862,12 +863,12 @@ var _ = Describe("Fly CLI", func() {
 		Context("when passing the pipeline argument", func() {
 			BeforeEach(func() {
 				cmdArgs = append(cmdArgs, "-p")
-				cmdArgs = append(cmdArgs, "some-pipeline")
+				cmdArgs = append(cmdArgs, "some-pipeline/branch:master")
 
 				expectedURL = "/api/v1/teams/main/pipelines/some-pipeline/builds"
-				queryParams = "limit=50"
+				queryParams = []string{"instance_vars=%7B%22branch%22%3A%22master%22%7D", "limit=50"}
```

Comment:
Not a big deal, but I'm wondering if we could encode the `instance_vars` differently in the query parameters, since URL encoding raw JSON isn't all that easy to read

What if we were to encode it as we do in the `PipelineFlag`...for instance: `?instance_vars=branch:master`

We could still URL encode the var names + values if need be. e.g. we could have `instance_vars=foo:bar%3Abaz` to represent `{"foo": "bar:baz"}`

This might be totally unnecessary, what do you think?

Reply:
I tried that, but the query is automatically encode in `go-concourse/concourse/internal` and I'm not sure toying with it is a good idea!

But I'm glad you raised this point because I wanted to start a discussion about using the pipeline's Id instead of the alternate key... Or at least allowing both in the api.

I've spent _some_ time in Elm and learned that the nature of the instance vars (`Dict String String`) can't be used in another `Dict`'s key, so the `teamName/PipelineName/InstanceVars` combination is a dead-end... But the pipeline's Id is always available and could be used instead!

In other words, I think we should allow both:
* `/api/v1/teams/:team_name/pipelines/:pipeline_name?instance_vars=url_encode_json`
* `/api/v1/teams/:team_name/pipelines/:pipeline_id`

Or for backward compatibility `/api/v1/teams/:team_name/pipelines/:pipeline_name_or_id?instance_vars=url_encode_json`, where the `pipeline_name_or_id` is just a check on the parameter's type or some other simple mechanism.

Any thoughts?

Reply:
*If* we want to support querying pipelines in the API by their id, it might make more sense for the endpoint to be `/api/v1/pipelines/:pipeline_id`, since the ID is unique across teams. But I'm trying to understand if that's necessary - I think the dead-end you're running into can be avoided by converting the `Dict String String` into a datatype that is a `comparable` (like a `List (String, String)` via `Dict.toList`, possibly also sorting the keys if necessary).

Also, side note - I was under the impression that instance vars could be any type (not just `string`), and could even have nested objects. e.g.

```yaml
set_pipeline: something
file: some-file.yml
instance_vars:
  foo: {bar: baz}
  qux: [1.2.3, 4.5.6]
```

@vito is that true?

Haven't looked through all of the code - wondering if this is something you've considered

Reply:
@aoldershaw Hmm... that's true, though now that you mention it I don't know if we've even considered the potential complexity from that in other places like the UI. I suppose the simplest thing may be to just JSON encode the non-string values everywhere we show them in the UI.

I like the idea of adding a `/api/v1/pipelines/:pipeline_id` endpoint. That will be more future-proof, especially if big dramatic things happen like [teams being replaced by projects](https://github.com/vito/rfcs/blob/5679ee86b4e810a8ae8e00dcb21aca2ad745103e/032-projects/proposal.md#can-projects-replace-teams), and potentially allowing projects to be nested within each other. So any change that gets us away from these hierarchical paths feels like a step in the right direction.

Reply:
The approach I took for the [`across` step](https://github.com/concourse/concourse/pull/5887) was to go 1 level deep into objects - so, `foo: {bar: 1, baz: 2}` would be rendered as two "vars" in the combination headers: `foo.bar: 1`, `foo.baz: 2`. Similarly for lists: `foo: [1, 2]` -> `foo[0]: 1`, `foo[1]: 2`. Anything deeper just gets encoded as raw JSON: `foo: {bar: {baz: 1}}` -> `foo.bar: {"baz": 1}`.

I figured that it wouldn't be uncommon to have vars 1 level deep, so we should optimize for that case. You can see an example of this in the screenshots in the PR description (`upgrade_path.from`, `upgrade_path.to`)

Reply:
Yeah, nested objects in instance vars make sense.. How would it looks like in fly commands? I think any kind of raw JSON would be unusable... Maybe combining all headers? eg.: `fly get-pipeline -p pipeline-name/foo.bar.baz:1,foo.bar.qux:2,bar[0]:1,bar[1]:2`?

Reply:
Raw JSON would be a bit annoying with all the quoting, but perhaps we could use a dumbed-down form of YAML (basically just JSON without the quotes):

`fly get-pipeline -p pipeline-name/foo:{bar:{baz:1,qux:2}},bar:[1,2]`

To parse it, you could probably even just take the whole string `foo:{bar:{baz:1,qux:2}},bar:[1,2]`, wrap it in `{...}`, and call `yaml.Unmarshal`.

Reply:
@aoldershaw That'll get weird with shells since `{}` is part of glob expansion. You'd have to quote the whole argument.

I wonder if we should enforce that instance vars are just regular identifiers. There's hidden complexity even in simple scalar values - if someone uses numbers, for example, we'd need to actually handle '2' as a number everywhere in the CLI. But what if someone specified them as a string like `"2"`? Would they need to quote it in the CLI? Will that not confuse the shell? e.g. `echo foo"bar"` just prints `foobar`.

Forcing them to be identifiers is a bit more limiting, but honestly it might be for the best - they're being used to identify the pipeline, and other vars can still be specified with arbitrary values.

Reply:
Though that kind of throws a wrench in some of my sketches, e.g. `instance_vars: {pr_number: 2}`.

Hmm...

Reply:
Could be valid identifiers or numbers? But then you can't use semver since that's a string.

Maybe we just loosen the restriction that valid identifiers can't start with a number, at least for instance vars?

Reply:
Yeah, that's tempting. My only hesitation is that I think instanced pipelines will be largely driven through Concourse itself with the `set_pipeline` step, and as a result the `fly` CLI interaction should be less than usual. I'm trying to be conscious not to let the CLI dictate other parts of the design *too* much, since it's not really the primary usage pattern; users typically shouldn't be manually archiving, destroying, exposing, or hiding individual instances. (I can see the argument for pausing and unpausing, even though that can be done in the UI.)

The most common command would probably be `fly intercept`, but I think it'd be better if that just took a container handle copied from the UI anyway. There's also `fly intercept --url` already.

I guess we could also accept a pipeline ID which can be copied from the UI somehow, or discovered through `fly pipelines`. I'm not sure users would want to bother typing it all out in the first place.

Don't want to just toss in a bunch of uncertainty and then bail on the convo, so: at this point I'm leaning towards keeping arbitrary values and just coming up with some sort of syntax for specifying them, which is what y'all were already doing. My one suggestion would be to not worry too much about arbitrary complex values, and just support accepting a pipeline ID in those cases. We'll see how users use `fly` with instances - it might just not be a big deal.

Reply:
Just to be clear... Instance vars are vars like all others, right? 

We have a jsonb column in db, and I can confirm that using more complex json is feasible and works fine (with a minor update pending...) for `set_pipeline` step.

Now about the fly commands, I think flattening/unflattening the json into dotted notation (eg.:  `pipeline-name/foo.bar.baz:1,foo.bar.qux:2,bar.0:1,bar.1:2`) works best, and it's as-is for simple instance vars (eg.:  `pipeline-name/branch:issue/5432`). 

If we want something simpler, then I'd suggest using the pipeline's id.

Reply:
@mouellet That's right. The dot notation seems like a good idea, since it avoids curly braces. My only concern then would be whether those '2' values are parsed as strings or numbers, since it'll have to match the pipeline's var values. Maybe we should just parse the values as YAML? If someone had a string that contained a number, then they could do this to disambiguate:

```
fly ... -p 'pipeline-name/foo.bar.baz:"1"'
```

Single quotes being significant here so the shell doesn't eat the double quotes.